### PR TITLE
[MetaSchedule] Add MultiLevelTilingTensorCore rule for auto-tensorization on CUDA

### PR DESCRIPTION
**Pull request by vinx13**
Co-authored-by: Bohan Hou \<32121147+spectrometerHBH@users.noreply.github.com\>
Co-authored-by: Hongyi Jin \<3231950289@qq.com\>

cc @junrushao1994 @masahi @spectrometerHBH @jinhongyii

### CODE REVIEW

**Review by masahi — COMMENTED** [review]
I haven't started looking at `multi_level_tiling_tensor_core.cc` yet.

How about providing an integration test to demonstrate that auto-tensorization on cuda works now?

File: src/meta_schedule/postproc/rewrite_tensorize.cc [hunk 1]
```diff
@@ -35,26 +35,24 @@ void CollectTensorizationJobs(
...
-          });
-        }
+        jobs->emplace_back(block_name, func_name, [sch, intrin_name](tir::BlockRV block) {
+          try {
+            sch->Tensorize(block, intrin_name.value());
+          } catch (const std::exception& e) {
+            LOG(WARNING) << "Tensorize failed with error " << e.what();
+          }
+        });
+      } else if (block_name.find("init") && vectorize_init_loop) {
```
[the remark this answers is not in the document]
Comment (vinx13) [reply src/meta_schedule/postproc/rewrite_tensorize.cc:48]: In rewrite reduction block, `tir::attr::meta_schedule_auto_tensorize` will be removed from the init block by default, unless the original reduction block is annotated with `tir::attr::meta_schedule_auto_tensorize_init`

File: src/meta_schedule/schedule_rule/multi_level_tiling_tensor_core.cc [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    f_transform_buffer_layout(tir::BufferIndexType::kWrite, i);
+  }
+
+  // Transform the layout of current block and reindex blocks
+  state->sch->TransformBlockLayout(state->tensor_core_reindex_store, index_map);
+  state->sch->TransformBlockLayout(state->tensor_core_reindex_A, index_map);
+  state->sch->TransformBlockLayout(state->tensor_core_reindex_B, index_map);
+  state->sch->TransformBlockLayout(state->block_rv, index_map);
+
+  return tir::TileWithTensorIntrin(state->sch, state->block_rv, intrin_group.compute_intrin);
```
Comment (masahi) [inline src/meta_schedule/schedule_rule/multi_level_tiling_tensor_core.cc:327]: you meant `intrin_name` here?

**Review by masahi — APPROVED** [review]

File: python/tvm/meta_schedule/testing/schedule_rule.py [hunk 3]
```diff
@@ -110,6 +112,32 @@ def multi_level_tiling(target: Target) -> ScheduleRule:
     raise NotImplementedError(f"{target.kind.name} is not supported")
 
 
+def multi_level_tiling_tensor_core(
+    target: Target, scope="shared", in_dtype="float16", out_dtype="float32", trans_b=False
```
[the remark this answers is not in the document]
Comment (vinx13) [reply python/tvm/meta_schedule/testing/schedule_rule.py]: it's write scope here but I think we also need a read scope param to support different read scopes

File: src/meta_schedule/schedule_rule/multi_level_tiling_tensor_core.cc (hunk 2 above)
Comment (vinx13) [reply src/meta_schedule/schedule_rule/multi_level_tiling_tensor_core.cc:338]: `intrin_group.compute_intrin` is intended

**Pull request merged**